pypit/par/pypitpar.py

```diff
@@ -0,0 +1,2486 @@
+        kwargs = {}
+        for pk in parkeys:
+            kwargs[pk] = cfg[pk] if pk in k else None
+        return cls(**kwargs)
+
+    @staticmethod
+    def valid_methods():
+        """
+        Return the valid methods to use for tilt tracing.
+        """
+        return [ 'pca', 'spca', 'spline', 'interp', 'perp', 'zero' ]
+
+    def validate(self):
+        pass
+
+
+class TraceObjectsPar(ParSet):
+    """
+    Parameters specific to PypIts object tracing algorithm
+    """
+    def __init__(self, function=None, order=None, find=None, nsmooth=None, xedge=None, method=None,
+                 params=None):
+
+        # Grab the parameter names and values from the function
+        # arguments
+        args, _, _, values = inspect.getargvalues(inspect.currentframe())
+        pars = dict([(k,values[k]) for k in args[1:]])      # "1:" to skip 'self'
+
+        # Initialize the other used specifications for this parameter
+        # set
+        defaults = dict.fromkeys(pars.keys())
+        options = dict.fromkeys(pars.keys())
+        dtypes = dict.fromkeys(pars.keys())
+        descr = dict.fromkeys(pars.keys())
+
+        # Fill out parameter specifications.  Only the values that are
+        # *not* None (i.e., the ones that are defined) need to be set
+        defaults['function'] = 'legendre'
+        options['function'] = TraceObjectsPar.valid_functions()
+        dtypes['function'] = str
+        descr['function'] = 'Function to use to trace the object in each slit.  ' \
+                            'Options are: {0}'.format(options['function'])
+
+        defaults['order'] = 2
+        dtypes['order'] = int
+        descr['order'] = 'Order of the function to use to fit the object trace in each slit'
+
+        defaults['find'] = 'standard'
+        options['find'] = TraceObjectsPar.valid_detection_algorithms()
+        dtypes['find'] = str
+        descr['find'] = 'Algorithm to use for finding objects.' \
+                        'Options are: {0}'.format(', '.join(options['find']))
+
+        defaults['nsmooth'] = 3
+        dtypes['nsmooth'] = [int, float]
+        descr['nsmooth'] = 'Parameter for Gaussian smoothing when find=nminima.'
+
+        defaults['xedge'] = 0.03
+        dtypes['xedge'] = float
+        descr['xedge'] = 'Ignore any objects within xedge of the edge of the slit'
+
+        defaults['method'] = 'pca'
+        options['method'] = TraceObjectsPar.valid_methods()
+        dtypes['method'] = str
+        descr['method'] = 'Method to use for tracing each object; only used with ARMED pipeline.' \
+                          '  Options are: {0}'.format(', '.join(options['method']))
+
+        defaults['params'] = [1, 0]
+        dtypes['params'] = [int, list]
+        descr['params'] = 'Parameters for the requested method.  For pca, params is a list ' \
+                          'containing the order of the polynomials that should be used to fit ' \
```

Comment:
It will be useful to add at the top of this module a short HOWTO
list on how to add a new parameter, or a new default, or a new 
Parset.

Reply:
I added comments on this in the file heading.